src/internal/storage/chunk/client.go

```diff
@@ -50,14 +50,41 @@ func (c *trackedClient) Create(ctx context.Context, md Metadata, chunkData []byt
 		panic("client must have a renewer to create chunks")
 	}
 	chunkID := Hash(chunkData)
+	needUpload, gen, err := c.beforeUpload(ctx, chunkID, md)
+	if err != nil {
+		return nil, err
+	}
+	if err := c.renewer.Add(ctx, chunkID); err != nil {
+		return nil, err
+	}
+	if !*needUpload {
+		return chunkID, nil
+	}
+	key := chunkKey(chunkID, *gen)
+	if err := c.store.Put(ctx, key, chunkData); err != nil {
+		return nil, err
+	}
+	if err := c.afterUpload(ctx, chunkID, *gen); err != nil {
+		return nil, err
+	}
+	return chunkID, nil
+}
+
+// beforeUpload checks the table in postgres to see if a chunk with chunkID already exists.
+// if err != nil then needUpload and gen will be nil
+// if *needUpload is false, then gen will be nil
+func (c *trackedClient) beforeUpload(ctx context.Context, chunkID ID, md Metadata) (needUpload *bool, gen *uint64, _ error) {
 	var pointsTo []string
 	for _, cid := range md.PointsTo {
 		pointsTo = append(pointsTo, cid.TrackerID())
 	}
 	chunkTID := chunkID.TrackerID()
-	var needUpload bool
-	var gen uint64
-	if err := dbutil.WithTx(ctx, c.db, func(tx *pachsql.Tx) error {
+	if err := dbutil.WithTx(ctx, c.db, func(tx *pachsql.Tx) (retErr error) {
+		// This function can be called multiple times.  In order to be sure
+		// that needsUpload and gen are set every run, we nil them here.
+		// This also protects this us from dbutil.WithTx dropping an error, if
+		// we only set these values before a `return nil`
+		needUpload, gen = nil, nil
```

Comment:
I would prefer us to just set the default value for the non-pointer types rather than having to dereference elsewhere. The pattern of returning default values or an error seems fine for non-pointer types and pointer types. Also, `protects this us from` -> `protects us from`. I'm not sure if we should really be coding around bugs in our own code, but ensuring the defaults are reset after a retry makes sense to me.

Reply:
The default value for needUpload is false, which is a little dangerous.  Panicking on a nil dereference is preferrable to silently skipping a chunk upload.  And I agree generally about not being defensive against our own code, but it would be particularly bad to have an error in this section.

Reply:
I don't think we should do this at all, even here where we want to be extra safe. I think we should stick to, and rely on, the pattern of nil error means the results are valid and a non-nil error means the results aren't. The decreased readability is worse than the possibility that we will catch the caller using the results with a non-nil error imo (since I think it is really unlikely to happen).

Reply:
> The decreased readability
This seems like a made up reason.  There are reasons not like this, but I don't see a case for readability being one of them.  The structure of the code and names are totally unchanged, there is a single additional assignment to nil, and an additional `*` when the variable is read.  It wouldn't impact the speed at which the average developer could traverse this area or understand what's going on.  I'd agree that it's inefficient to pass around a pointer to a smaller than pointer object, and perhaps not idiomatic Go.

What about changing this to a `skipUpload` boolean so that the default false is less dangerous, and removing the pointers?

Reply:
The not idiomatic Go part is why it is less readable imo. Using a pointer to a builtin type usually makes me have to consider how it is set (when the default isn't valid) and where it could be referenced, plus the additional step to actually access the value. I'm not convinced inverting the value is much safer because if the gen is incorrect then the upload will screw things up. I think we shouldn't be relying on setting the results in a particularly way to prevent the client from using them when they shouldn't. Setting the error should be enough.